src/components/views/rooms/RoomTileIcon.tsx

```diff
@@ -0,0 +1,148 @@
+
+    constructor(props: IProps) {
+        super(props);
+
+        this.state = {
+            icon: this.getIcon(),
+        };
+    }
+
+    private get isPublicRoom(): boolean {
+        const joinRules = this.props.room.currentState.getStateEvents("m.room.join_rules", "");
+        const joinRule = joinRules && joinRules.getContent().join_rule;
+        return joinRule === 'public';
+    }
+
+    public componentWillUnmount() {
+        this.isUnmounted = true;
+        if (this.isWatchingTimeline) this.props.room.off('Room.timeline', this.onRoomTimeline);
+        this.unsubscribePresence();
+    }
+
+    private unsubscribePresence() {
+        if (this.dmUser) {
+            this.dmUser.off('User.currentlyActive', this.onPresenceUpdate);
+            this.dmUser.off('User.presence', this.onPresenceUpdate);
+        }
+    }
+
+    private onRoomTimeline = (ev: MatrixEvent, room: Room) => {
+        if (this.isUnmounted) return;
+
+        // apparently these can happen?
+        if (!room) return;
+        if (this.props.room.roomId !== room.roomId) return;
+
+        if (ev.getType() === 'm.room.join_rules' || ev.getType() === 'm.room.member') {
+            this.setState({icon: this.getIcon()});
+        }
+    };
+
+    private onPresenceUpdate = () => {
+        if (this.isUnmounted) return;
+
+        let newIcon = this.getPresenceIcon();
+        if (newIcon !== this.state.icon) this.setState({icon: newIcon});
+    };
+
+    private getPresenceIcon(): Icon {
+        let newIcon = Icon.None;
+
+        const isOnline = this.dmUser.currentlyActive || this.dmUser.presence === 'online';
+        if (isOnline) {
+            newIcon = Icon.PresenceOnline;
+        } else if (this.dmUser.presence === 'offline') {
+            newIcon = Icon.PresenceOffline;
+        } else if (this.dmUser.presence === 'unavailable') {
+            newIcon = Icon.PresenceAway;
+        }
+
+        return newIcon;
+    }
+
+    private getIcon(): Icon {
+        let defaultIcon = Icon.None;
+        this.unsubscribePresence();
+        if (this.props.tag === DefaultTagID.DM && this.props.room.getJoinedMemberCount() === 2) {
+            // Track presence, if available
+            if (isPresenceEnabled()) {
+                const otherUserId = DMRoomMap.shared().getUserIdForRoomId(this.props.room.roomId);
+                if (otherUserId) {
+                    this.dmUser = MatrixClientPeg.get().getUser(otherUserId);
+                    if (this.dmUser) {
+                        this.dmUser.on('User.currentlyActive', this.onPresenceUpdate);
+                        this.dmUser.on('User.presence', this.onPresenceUpdate);
+                        defaultIcon = this.getPresenceIcon();
+                    }
+                }
+            }
+        } else {
+            // Track publicity
+            defaultIcon = this.isPublicRoom ? Icon.Globe : Icon.None;
+            this.props.room.on('Room.timeline', this.onRoomTimeline);
```

Comment:
onRoomTimeline calls this which registers a handler for onRoomTimeline which looks like a handler leak